### ISSUE TITLE
Unhandled rejection Error

### ISSUE THREAD

**Issue by anasyb** [balderdashy/waterline#1015]
Have been getting this error a bunch of times a day since a week or so. Have no idea why or from where is it originating in the async spaghetti. Spent couple of days trying to pinpoint it with no much luck.

The only difference I get is that the number of `[object Object]` listed varies, sometimes its one, sometimes its a bunch of them... or in this case its just 2.

Any ideas/hints guys on what would be the cuz of this and why the number of `[object Object]` is different each time?

```
Unhandled rejection Error: [object Object],[object Object]
    at Object.ensureErrorObject (/usr/local/lib/node_modules/sails/node_modules/waterline/node_modules/bluebird/js/main/util.js:226:20)
    at Promise._rejectCallback (/usr/local/lib/node_modules/sails/node_modules/waterline/node_modules/bluebird/js/main/promise.js:416:22)
    at Object.reject (/usr/local/lib/node_modules/sails/node_modules/waterline/node_modules/bluebird/js/main/promise.js:433:17)
    at /usr/local/lib/node_modules/sails/node_modules/waterline/lib/waterline/model/lib/defaultMethods/save.js:165:18
    at /usr/local/lib/node_modules/sails/node_modules/async/lib/async.js:454:17
    at /usr/local/lib/node_modules/sails/node_modules/async/lib/async.js:444:17
    at Array.forEach (native)
    at _each (/usr/local/lib/node_modules/sails/node_modules/async/lib/async.js:46:24)
    at Immediate.taskComplete (/usr/local/lib/node_modules/sails/node_modules/async/lib/async.js:443:13)
    at processImmediate [as _immediateCallback] (timers.js:369:17)
```

**Comment by dmarcelino** [comment]
Hi @anasyb, the only waterline related line I see in the stack trace is:
``` shell
    at /usr/local/lib/node_modules/sails/node_modules/waterline/lib/waterline/model/lib/defaultMethods/save.js:165:18
```

Which refers to [defaultMethods/save.js#L165](https://github.com/balderdashy/waterline/blob/master/lib/waterline/model/lib/defaultMethods/save.js#L165):
```javascript
        deferred.reject(failedTransactions);
```

I believe this code runs when you do `someModel.save(cb)`. Is it possible you're not processing/catching the result of one of those `.save()` operations?

**Comment by anasyb** [comment]
Thanks @dmarcelino, a quick find for all `.save(` revealed that this is the case... the error there is simple:
`Error: Trying to '.add()' an instance which already exists!`

I still get that error even after adding the callback. Why would it still throw an `Unhandled rejection Error: `? Shouldn't it just pass the error to my callback without throwing an `Unhandled rejection Error: `?

**Comment by dmarcelino** [comment]
Not sure I follow, can you share a code snippet?

**Comment by anasyb** [comment]
wow... this turned out to be weirder than I initially thought, the block of code looks something like:
```
User.create({...}).exec(function(err, u){
	Post.create({...}).exec(function(err, p){
		u.posts.add(p.id);
		u.save(console.log);
	});
	Post.create({...}).exec(function(err, p){
		u.posts.add(p.id);
		u.save(console.log); //this one is throwing an error
	});
})
```

**Console**
```
[ { type: 'insert',
    collection: 'user_posts__post_posts_post',
    criteria: 
     { post_posts_post: '555d104ec29c6701004cbf3b',
       user_posts: '555d104ec29c6701004cbf36' },
    values: 
     { post_posts_post: '555d104ec29c6701004cbf3b',
       user_posts: '555d104ec29c6701004cbf36' },
    err: [Error: Trying to '.add()' an instance which already exists!] },
  { type: 'insert',
    collection: 'user_posts__post_posts_post',
    criteria: 
     { post_posts_post: '555d104ec29c6701004cbf3c',
       user_posts: '555d104ec29c6701004cbf36' },
    values: 
     { post_posts_post: '555d104ec29c6701004cbf3c',
       user_posts: '555d104ec29c6701004cbf36' },
    err: [Error: Trying to '.add()' an instance which already exists!] } ]
Unhandled rejection Error: [object Object],[object Object]
    at Object.ensureErrorObject (/usr/local/lib/node_modules/sails/node_modules/waterline/node_modules/bluebird/js/main/util.js:226:20)
    at Promise._rejectCallback (/usr/local/lib/node_modules/sails/node_modules/waterline/node_modules/bluebird/js/main/promise.js:416:22)
    at Object.reject (/usr/local/lib/node_modules/sails/node_modules/waterline/node_modules/bluebird/js/main/promise.js:433:17)
    at /usr/local/lib/node_modules/sails/node_modules/waterline/lib/waterline/model/lib/defaultMethods/save.js:165:18
    at /usr/local/lib/node_modules/sails/node_modules/async/lib/async.js:454:17
    at /usr/local/lib/node_modules/sails/node_modules/async/lib/async.js:444:17
    at Array.forEach (native)
    at _each (/usr/local/lib/node_modules/sails/node_modules/async/lib/async.js:46:24)
    at Immediate.taskComplete (/usr/local/lib/node_modules/sails/node_modules/async/lib/async.js:443:13)
    at processImmediate [as _immediateCallback] (timers.js:369:17)
```

1. in any case, even if I was adding a duplicate record to the association, why is it throwing the `Unhandled rejection Error:` 
2. I don't understand how on earth adding a freshly created record could ever be a duplicate record?! Moreover, you can see in the console that `Error: Trying to '.add()' an instance which already exists!` is thrown twice with the same IDs... seems like waterline is trying to insert the associated record couple of times
3. I tried to duplicate the error on an isolated case but failed to do so?!
4. removing the first `Post.create` stops the error from in the second `Post.create`... no idea what does the first create has to do with the second.

**Comment by dmarcelino** [comment]
That code snippet is so naughty... not sure what exactly is causing the error to be honest, sounds like some sort of racing condition... but can you try:

``` javascript
User.create({...}).exec(function(err, u){
	Post.create({...}).exec(function(err, p){
		u.posts.add(p.id);
		Post.create({...}).exec(function(err, p){
		  u.posts.add(p.id);
		  u.save(console.log);
		});
	});
})
```

2. The thing is you are re-using `u` twice for adding the post association. I believe that in your second `u.save(console.log);` it actually retries to save the previous saved post because `u` still has it queued up to be saved.

**Comment by anasyb** [comment]
- I don't want to do it like the code you posted... cause these are completely independent. First could error, or could even not be called in the first place. To be honest, there are three of them :P not just two... but I commented the third one and still getting the error.
- So I tried to separate the two by a setTimeout on the second one.. it worked.
- this makes me think your idea is correct... probably its trying to save it again. I would consider this a bug though... `.save` followed by another `.save` shouldn't error!!

**Comment by dmarcelino** [comment]
The thing is that in terms of execution order your current code is essentially doing the equivalent of:
``` javascript
u.posts.add(id1);
u.posts.add(id2);
u.save(console.log);
u.save(console.log);  // oops
```
And I'm not sure this is valid, so I'm not sure this qualifies as a bug... In other words, I don't think you are expected to run `.save()` twice before the first operation concludes.

In my example, it is ignoring the `err` thrown by the first `Post.create({...})` so even if the first operation fails, the second would still succeed. You could also use `async` for the posts creation and, in the end, call the `.save()` operation.

**Comment by anasyb** [comment]
- I understood the problem and proposed solution: basically insure to call `save` just once. Whether by cascading the creates, or using async. This assumption is problematic when you have several parts of the code trying to update/initialize parts of the newly created fella. If its confirmed waterline doesn't support this I suggest:
  - having a feature request for it
  - warning devs against calling it twice on the re-used the moment they call it 
- If the above isn't true (i.e. its ok to call twice), you could argue that two consecutive calls to `save()` might attempt at inserting the record twice, and therefore the second `save` would throw a duplicate keys err like when you attempt to insert a record twice. I would argue otherwise. But regardless of this or that, the `Unhandled rejection Error:` is a bug... inserting twice never showed such error in the console!!

**Comment by dmarcelino** [comment]
Just to be clear, I don't think the problem is calling `.save()` *twice*, I think the problem is calling `.save()` *twice concurrently on the same object*. I bet `u.save(function(){ /*...*/ u.save(cb); })` would work, didn't test but your earlier `setTimeout()` experience suggests it.

**Comment by anasyb** [comment]
Thanks you're absolutely correct about that, indeed calling it twice isn't a problem... my new code now calls it three times at some cases without an issue. I am editing my comment above... which still applies.
But I think allowing it to be called twice, but not concurrently, is tricky... so I definitely vote for a feature request there.
I don't know if the unhandled rejection error is just appearing in this very specific case, or is it being fired always when there is an error with the save.

**Comment by dmarcelino** [comment]
For reference, PR #1048 should fix this.

**Comment by Nitemaeric** [comment]
I've come up with the same error.

In my case it's because of the following format, _(all names are fake names that represent their role)_:

```javascript
Model.update(id, {
    collection: array_of_ids
});
```

This gives me the following error, which it did not give before:

```
Unhandled rejection Error (E_UNKNOWN) :: Encountered an unexpected error
    at new WLError (/usr/local/lib/node_modules/sails/node_modules/waterline/lib/waterline/error/WLError.js:26:15)
    at duckType (/usr/local/lib/node_modules/sails/node_modules/waterline/lib/waterline/error/index.js:67:10)
    at errorify (/usr/local/lib/node_modules/sails/node_modules/waterline/lib/waterline/error/index.js:39:10)
    at wrappedCallback (/usr/local/lib/node_modules/sails/node_modules/waterline/lib/waterline/utils/normalize.js:325:15)
    at callback.error (/usr/local/lib/node_modules/sails/node_modules/waterline/node_modules/switchback/lib/normalize.js:42:31)
    at _switch (/usr/local/lib/node_modules/sails/node_modules/waterline/node_modules/switchback/lib/factory.js:46:28)
    at /usr/local/lib/node_modules/sails/node_modules/waterline/lib/waterline/query/dql/update.js:51:22
    at /usr/local/lib/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:396:17
    at /usr/local/lib/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:183:21
    at /usr/local/lib/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:393:21
    at /usr/local/lib/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:768:34
    at /usr/local/lib/node_modules/sails/node_modules/waterline/lib/waterline/query/validate.js:70:22
    at /usr/local/lib/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:396:17
    at /usr/local/lib/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:183:21
    at /usr/local/lib/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:393:21
    at /usr/local/lib/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:768:34
```

The only change I've recently done is updating my Node.js version to 0.12.7. As before 0.12.6 it was giving binding errors for bcrypt.

**Comment by sailsbot** [comment]
Thanks for posting, @anasyb.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
